### PR TITLE
Display Buffer Enhancement

### PR DESCRIPTION
**Pull request by SenexCrenshaw**
## Description:
Added a display buffer middleman.

The display class will use it with 
`public display_buffer_ex::DisplayBufferEx`
instead of
`public display::DisplayBuffer`

The initializer is the one piece that needs to be added. 
`this->init_buffer(this->width_, this->height_,this->eightbitcolor_ ? ColorBit::COLOR_BIT_332 : ColorBit::COLOR_BIT_565);`

This will initialize the buffer with 100x200x1 byte since 332 needs 8 bits. 565 will store it as 2 bytes.  Color will be converted to 332 or 565 automatically when a buffer update is requested (draw_absolute_pixel_internal). 

the conversion allows the display developer to not worry about this on write unless a conversion is warranted. 
If init was set with ColorBit::COLOR_BIT_565 then no work is needed on write
`this->write_array(this->buffer_, this->get_buffer_length());` 

This is specific to 332 and 565 but I believe it is easy to add any others if folks want to.

My hope is the color changes make their way in. I will port this to use that and then I can change my display driver. To use this!
#1381 

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [X ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by glmnet** [comment]
Not really what I had in mind.
I guess inheritance like this is not going to work, I want some mechanism to make screen buffer independent from the driver, if we use inheritance, then when the driver (most specific display class) inherits from a buffer, then that is the buffer, and no more changes could be made.

I don't know how to do it right away, would have to mock up some code, or you can do it if you want to experiment with me.

I'd like the yaml to look like:

```yaml
display:
  - platform: any_display
    name: display_1
    buffer: color_8bpp # color eight bits per pixel, like 553
    ...
```

other buffer types, like said on other posts, could be monochromatic (to be used with two color displays, but not only those), 4 bits, 16, etc, then paletted like 16 colors which is going to be 4bpp, etc.

each buffer class should do just that, i.e. not options in the runtime like `this->eightbitcolor_ ? ...` the buffer itself will be a specific type already.

I hope you tune in here

**Comment by SenexCrenshaw** [comment]
> Not really what I had in mind.
> I guess inheritance like this is not going to work, I want some mechanism to make screen buffer independent from the driver, if we use inheritance, then when the driver (most specific display class) inherits from a buffer, then that is the buffer, and no more changes could be made.
> 
> I don't know how to do it right away, would have to mock up some code, or you can do it if you want to experiment with me.
> 
> I'd like the yaml to look like:
> 
> ```yaml
> display:
>   - platform: any_display
>     name: display_1
>     buffer: color_8bpp # color eight bits per pixel, like 553
>     ...
> ```
> 
> other buffer types, like said on other posts, could be monochromatic (to be used with two color displays, but not only those), 4 bits, 16, etc, then paletted like 16 colors which is going to be 4bpp, etc.
> 
> each buffer class should do just that, i.e. not options in the runtime like `this->eightbitcolor_ ? ...` the buffer itself will be a specific type already.
> 
> I hope you tune in here
I think we are aligning. Do you envision the each buffer type (8bit,2bit) be a separate class? I was thinking of a more generic one that could be configured with ENUMs but it sounds it might be best to maybe have a base buffer class that the other more specific ones can work off of. Keeping the API consistent. 
If we went all generic something like this would work for my case:
```
displaybuffer_332:
	id: buffecolor_8bpp_332_565
	color_bit_size: 8	
	write_colorformat: RGB332
	read_colorformat: RGB565
	width: 160
	height: 120
display:
  - platform: any_display
    name: display_1
    buffer_id: buffecolor_8bpp_332_565
```
If not we could do:
```
displaybuffer_332_565:
	id: buffecolor_8bpp_332_565
	color_bit_size: 8	
	width: 160
	height: 120
```

the constructor for the display driver would have the buffer component passed to it. I think on the buffer side we wont need much besides an init/setup with the this->buffer_ from the displaybuffer being passed to it. After that just a write and a read from the buffer and we should be good.

Thoughts?

**Comment by glmnet** [comment]
yes, passing the buffer via id will work better as you can setup the buffer properly, buffer setup might still receive parameters, 
i.e. the driver can init the buffer with size information, so size information will not be needed there (it will be redundant at some point, unless something more complex goes on, e.g. scaling) 
the display `buffer_id` could be optional too, so the display itself can decide which buffer type to use, it can even default to different types based on platform (esp8266 / esp32 so far) also we can keep backwards compatibility at least in the yaml.

having said that, having a `buffer_id` which will require you to fill a few more yaml lines might be a bit ugly on yaml side, we could also have a `buffer_type` exclusive optional which will create the buffer with default values (I envision buffers with no options) Under the hood it could be very much the same though.

the buffer should accept the `draw_pixel_absolute` thing, so it will be color format independent (the method accepts `Color`)
Then the buffer should transform it without checking any parameter, if you want a buffer for RGB332 then the buffer converts color to 332, or if you want 8 pixels per bit, the buffer updates one bit in the bitmap, I would like this to happen on the buffer subclass, not in color.h file. If you check buffer options on the `draw_pixel_...` method, then having different class types stops making sense.

then the driver should be able to consume the buffer, may be some kind of multiple inheritance can be used, I don't know all alternatives available in C++ for this kind of stuff, but a buffer can provide a `get_pixel_*` set of optional methods, e.g.
`get_pixel_on_off()` for binary display, `get_pixel_gray8()`, `get_pixel_rgb556()` etc.
Not all buffers would need to implement all those, it does not make sense for a rgb332 buffer to implement `get_on_off()` (the other way does make sense as you might want to use a color display with only two colors)

---
this is all opinionated stuff, what I'm proposing can be very bad design ideas, I'm not trying to convince my ideas are better, I just want to come to a good design before we move on.

**Comment by SenexCrenshaw** [comment]
I agree and enjoy the conversation.
For my use case ( user can select 16/8bit )
I would create 2 new classes called buffer_332, buffer_565. They would inherit from the base class. 
Since we are using a base class I would do something like this:
```
Buffer_332 buffer_332;
Buffer_565 buffer_565;
ColorBuffer *color_buffer = nullptr;
if ( eightbit){
  color_buffer =&buffer_332;
} else{
  color_buffer =&buffer_565;
}
```

On setup of the display driver I would choose which one to use and then init the buffer passing the display buffer (this->buffer_) to it.
`void color_buffer->init_buffer(int width, int height,uint8_t *buffer);`
It would have a few functions in the base:
```
void set_pixel(int x, int y, Color color);
size_t get_buffer_length();
```
Then just use color_buffer in my driver code. 

If we wanted we could just put a bunch of virtual get_pixels_* in that base class and let the buffer author decide to override them or not if they inherit from that base class. 
Either way in my buffer_332 I would have
```
get_pixel_rgb332(); //really not needed for my purpose 
get_pixel_rgb565();
```
I don't think I would use get_pixel_gray8()  but if we felt all color displays had to implement that we could force it. 

Thoughts?

**Comment by glmnet** [comment]
> ```
> Buffer_332 buffer_332;
> Buffer_565 buffer_565;
> ColorBuffer *color_buffer = nullptr;
> if ( eightbit){
>   color_buffer =&buffer_332;
> } else{
>   color_buffer =&buffer_565;
> }
> ```

I expect this to be solved in the Python side, generated code will create the desired buffer either 332 or 565 and pass it somehow to the driver, like, e.g. fonts are passed.





> Either way in my buffer_332 I would have
> 
> ```
> get_pixel_rgb332(); //really not needed for my purpose 
> get_pixel_rgb565();
> ```
> 
> I don't think I would use get_pixel_gray8() but

`get_pixel_*` makes sense for the drivers only:
there are drivers which need a pixel_on_off() method, others a rgb565() etc, if there is no driver which requires a rgb332() (which I believe there are none) then it does not make sense to create this function.

**Comment by SenexCrenshaw** [comment]
I think I know what you are looking for. I'll mock something up tomorrow morning. This is really helping me learn more about esphome which I'm greatly enjoying!

**Comment by SenexCrenshaw** [comment]
Got a little busy today but i have the bones down
Here is a preview though:
```
bufferex_332:
  id: bufferex_332_1

- platform: st7735
    model: 'INITR_BLACKTAB'
    bufferex_id: bufferex_332_1
    cs_pin: GPIO5
    dc_pin: GPIO16
    reset_pin: GPIO23
    rotation: 0
    devicewidth: 128
    deviceheight: 160
    colstart: 0
    rowstart: 0
    lambda: |-
      it.rectangle(0, 0, it.get_width(), it.get_height());
```

My driver code doesnt need to change if I change the buffer backend. Pretty cool! So I can create 1 bit (2color) buffer and swap that in and it just works.  Hopefully this is what you are looking for. 

Oh yeah. 1 bit 2 color only needs 2400bits for my entire display.

**Comment by SenexCrenshaw** [comment]
doh. must have hit the wrong button. didnt mean to close!

**Comment by SenexCrenshaw** [comment]
First I am not married to the name of these classes. 
bufferex_base - Base class with virtuals for the functions we want the buffers to have. bufferex_base is passed to the display driver.
bufferex_332 - Derived class of bufferex_base. handles writing 332 into an 8bit buffer and any color our conversions.
bufferex_565 - Derived class of bufferex_base. handles writing 565 into an 16bit buffer and any color our conversions
.
I only implemented to_332/to_565 in the buffers for now. I wanted to get feedback on this. BTW works great on my display driver #1066 that i would port once we have something here we like. 

bufferex_332/bufferex_565 can be swapped with this in my driver yaml stanza:
bufferex_id: bufferex_332_1

The python code takes care of the backend code for which buffer is allocated.

Works like a charm. 

So I think I want a way to override the size of the buffer instead of width*height*(int8/16) . I want a 1bit 2 color buffer so i need to be able to provide my own size. 

Thoughts on all of this?

**Comment by glmnet** [comment]
Sounds good I'll take a closer look when I have some time. 
Meanwhile I merged your display so you can rebase this and we can start viewing how display drivers are going to be changed. 
I'd say this PR also include the ST display changes.

**Comment by SenexCrenshaw** [comment]
> Sounds good I'll take a closer look when I have some time.
> Meanwhile I merged your display so you can rebase this and we can start viewing how display drivers are going to be changed.
> I'd say this PR also include the ST display changes.

Sounds good. Ill submit another PR

**Comment by SenexCrenshaw** [comment]
I added in a bufferex_1bit_2color buffer to show how we can create custom buffers and initialize them with their own settings. This one happens to store 1 bit per color which are set in the config or default to black/white. My whole display takes 2560 bit for this. 

I dont care if that makes its way into esphome. its more of a demo if anything, it could come in handy and maybe a good start to palleted buffers.

**Comment by glmnet** [comment]
Actually it could replace the buffers used in binary displays, why not?

**Comment by SenexCrenshaw** [comment]
> Actually it could replace the buffers used in binary displays, why not?

You mean like epaper displays? I think any buffer could be created now to handle these types of situations.

I want to make a 2 bit one for a UI display I have. I get 4 colors to play with and can change them on the fly. So I can do things like blink a circle (change colors) with ease. A color is tied to the geo/text your are creating so just changing the color through lambda or backend code is simple since the color code is stored in a simple variable and not buried in a large buffer. 
id(bufferex1)->set_color1(Color(255,0,0));
sleep 1 second
id(bufferex1)->set_color1(Color(0,0,0));

File: esphome/components/bufferex_base/bufferex_base.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ public:
+  uint8_t *buffer_{nullptr};
+  virtual void init_buffer_with_size(int width, int height, uint8_t pixel_bit_size);
+  virtual void init_buffer(int width, int height);
+  virtual void set_pixel(int x, int y, Color color) = 0;
+  uint8_t get_pixel_bit_size() { return pixel_bit_size_; };
+  size_t get_buffer_length();
+
+  // 332
+  virtual uint8_t get_pixel_to_332(int x, int y, Color::ColorOrder color_order) = 0;
```
Comment (glmnet) [inline esphome/components/bufferex_base/bufferex_base.h:16]: where is this method used?

File: esphome/components/bufferex_base/bufferex_base.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  // uint8_t get_pixel_rgb332();
+  // uint8_t get_pixel_gbr332();
+  // uint8_t get_pixel_grb332();
+
+  // 565
+  virtual uint16_t get_pixel_to_565(int x, int y, Color::ColorOrder color_order) = 0;
```
Comment (glmnet) [inline esphome/components/bufferex_base/bufferex_base.h:23]: where this method will be used? and which parameters will be used? (which different color_order)
I would highly lean to have a unique color order and have the color being reordered in the display driver, but to make a good choice we have to check all the required color orders so far
This is key for me
So far I know we have 3 displays:
B&W, like, SSD1306, epaper and the MAX7219 in matrix mode, etc.
565 RGB like the ST7789, ILI9341
and 4 bit grayscale

so in order to see what really needs to be implemented we need to look at this carefully

File: esphome/components/bufferex_565/bufferex_565.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+void Bufferex565::init_buffer(int width, int height) { BufferexBase::init_buffer_with_size(width, height, 16); }
+
+void HOT Bufferex565::set_pixel(int x, int y, Color color) {
+  if (x >= this->width_ || x < 0 || y >= this->height_ || y < 0)
+    return;
+
+  const uint16_t color565 = color.to_565();
+  uint16_t pos = get_pixel_buffer_position_(x, y);
+  this->buffer_[pos++] = (color565 >> 8) & 0xff;
+  this->buffer_[pos] = color565 & 0xff;
```
Comment (glmnet) [inline esphome/components/bufferex_565/bufferex_565.cpp:16]: may be buffer should be created in the most specific class, so you can actually have an array of `uint_16t` here and simplify access to pixels

**Review by glmnet — COMMENTED** [review]
In general what I do not like about this is that it creates a lot of directories in the components folder, I don't know which would be good solution though, do sb have an idea?

Also general naming my be reviewed
are we still support display buffer non ex?

File: esphome/components/bufferex_1bit_2color/__init__.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    cv.GenerateID(): cv.declare_id(bufferex_1bit_2color),
+    cv.Optional(CONF_COLOR_ON): cv.use_id(color),
+    cv.Optional(CONF_COLOR_OFF): cv.use_id(color),
+})
+
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    if CONF_COLOR_ON in config:
+        con = yield cg.get_variable(config[CONF_COLOR_ON])
```
[the remark this answers is not in the document]
Comment (jesserockz) [reply esphome/components/bufferex_1bit_2color/__init__.py:28]: Yeah it doesn't really make sense especially in the python code which should be as readable as possible.

**Comment by SenexCrenshaw** [comment]
This code was just meant to be a POC or example of this on going discussion. I think maybe we look at this from another perspective. As a front end UI developer for an IoT device I need basic primitives (boxes/circles..) with a color associated with it. I also might want a picture primitive for a nice background or fancy icon (layers?)

To accomplish this we can create a palette component that has an array of colors. I can draw a box and use palette[2] for its color.  For my pictures, I would format them in whatever way needed to optimize for space and color(332/556).

Essentially the bufferex (whatever name) or maybe just displaybuffer.. would know about a palette and also be able to store a bunch of mini binary buffers for whatever color photo/icon needed. A getpixel function would do the work on returning a color to the display driver.

Make sense? so essentially we are not going to store a color for every pixel. We will be storing an index number of a color from a palette for every pixel except. The exception would be for the the binary pictures involved. Those would only be the size needed to store the picture. As a display developer I dont care anymore about color conversion to save space or whatever. Give me a pixel in the format I want and I will draw it. If the end user is doing something crazy with their UI it will be on them to make sure they dont overcommit available memory.

With a tying a primitive to a index of a palette we can add effects too.  Blinking/fading would be easy. We can also swap palettes in an out with almost no CPU cycles to do that. The next refresh will take care of that.

I think this approach in general might make some lives easier and will save a ton of RAM. 4 colors might be enough for most applications and the palette storage piece will be smart enough to know how to optimally store what it needs in RAM. Can I sneak in some full color/332/565 pictures, sure and it will only need as much space as the picture itself. 

The big question is what is the use case? Its maybe redesigning UI development in esphome. It allows 8266 users to have a lot more RAM to play with. It gives us some flexibility and opens up the possibility to add in some cool effects. As these IoT devices become more powerful we will be able to do more things from the frontend point of view as well.  

Just my 2 cents.  I am not married to any of this or saying we need to do anything. I do appreciate the code comments as well. This was just a quick example but I do need better habits!

**Comment by glmnet** [comment]
I agree with you in the comments. I really don’t understand what’s the big point in these statements. 

I also believe the motivation is related to memory and performance, otherwise we could have a ARGB24 buffer and let this whole thing alone. 

Also allowing using a color display with restrictions in esp8266 while allowing a less restricted environment in other more powerful platforms. 

So your code is already doing some of this, it’s going well I believe.

**Comment by SenexCrenshaw** [comment]
OMG I clearly had too much coffee on that post! I have an idea on what to do and will hope to put something together this week.

File: esphome/components/bufferex_1bit_2color/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SenexCrenshaw) [reply esphome/components/bufferex_1bit_2color/__init__.py:28]: I agree! I was just prototyping this out for discussion but always need better habits. ill adjust

File: esphome/components/bufferex_base/bufferex_base.h (hunk 1 above)
Comment (SenexCrenshaw) [reply esphome/components/bufferex_base/bufferex_base.h:16]: We dont need this. I dont think there will be any need for an 8bit display.

File: esphome/components/st7735/st7735.cpp [hunk 5]
```diff
@@ -441,18 +423,16 @@ void HOT ST7735::write_display_data_() {
...
-      for (int index = 0; index < this->get_width_internal(); ++index) {
-        auto color = Color(this->buffer_[index + line], Color::ColorOrder::COLOR_ORDER_RGB,
-                           Color::ColorBitness::COLOR_BITNESS_332, true)
-                         .to_565();
+  if (this->bufferex_base_->get_pixel_bit_size() == 16) {  // 16bit just write. I am a 565 display
+    this->write_array(this->bufferex_base_->buffer_, this->bufferex_base_->get_buffer_length());
+  } else {
+    for (int y = 0; y < this->get_height_internal(); ++y) {
+      for (int x = 0; x < this->get_width_internal(); ++x) {
+        auto color = this->bufferex_base_->get_pixel_to_565(x, y, Color::ColorOrder::COLOR_ORDER_RGB);
```
[the remark this answers is not in the document]
Comment (SenexCrenshaw) [reply esphome/components/st7735/st7735.cpp:431]: I was going more for flexibility if someone wanted a "unique" buffer size but I think an ENUM for each buffer type is much better. I added that in

File: esphome/components/bufferex_base/bufferex_base.h (hunk 2 above)
Comment (SenexCrenshaw) [reply esphome/components/bufferex_base/bufferex_base.h:23]: I agree we should keep color conversions like ordering out of the buffer. I removed that

File: esphome/components/bufferex_565/bufferex_565.cpp (hunk 3 above)
Comment (SenexCrenshaw) [reply esphome/components/bufferex_565/bufferex_565.cpp:16]: I thought about that but couldnt do 
``` this->write_array(this->bufferex_base_->buffer_, this->bufferex_base_->get_buffer_length());```
with a 16bit buffer size. BUT to make things a lot cleaner (looking at least) I redid the PR moving the buffer like you mentioned and I really like how the code looks.

**Comment by SenexCrenshaw** [comment]
@glmnet I made the changes you were looking for. The code looks clean to use and it natively stores the buffer data. So 565 gets uint16_t and doesnt do any conversion coming out.

@kbx81 What are your thoughts? overkill? Your PR #1405 for 4 bit buffer could be converted to one of these buffers and then everyone could use it. 

Let me know if you guys think this is valuable. Im not advocating for this but I think as we add in more displays that a way to store the buffer data with in a smaller footprint would be welcomed and at the very least we will handle things like 888 to 565 in a consistent way.

File: esphome/components/bufferex_base/bufferex_base.h (hunk 2 above)
Comment (kbx81) [reply esphome/components/bufferex_base/bufferex_base.h]: I've got quite a few of the various displays we support now (I think almost all of the OLEDs, just none of the e-paper ones) so I could help test and would be happy to work to update the displays I've worked on already. 😄

**Comment by glmnet** [comment]
I like it, good job 👍

I'd like people to give feedback too.

File: tests/test1.yaml [hunk 6]
```diff
@@ -1708,6 +1717,7 @@ display:
       it.rectangle(0, 0, it.get_width(), it.get_height());
   - platform: st7735
     model: 'INITR_BLACKTAB'
+    bufferex_id: bufferex_332_1
```
Comment (glmnet) [inline tests/test1.yaml:1720]: It'd be nice if this were optional, the display python code would check if it's not defined then create a default buffer (the type could depend on the platform) this way we can add this feature to other displays without introducing breaking changes.

**Comment by kbx81** [comment]
Ok...I finally spent some time this evening to read through all of this and I think I have a handle on it. I'm probably going to end up reiterating some of what has already been discussed, so please for give me if any of it is repetitive.

The primary need that I see is that we need to provide a way to use these large/color/high-resolution displays on memory-constrained devices.

I think that this accomplishes that in a reasonable way. No matter what, we are adding complexity here and there's naturally going to be a trade-off for terms of simplicity. We are also trading performance for memory.

Overall I think that this is meets those needs -- the additional flexibility it provides is great, even if it adds a little complexity. I am on board with the change and I think it's looking great!

This may be a somewhat different issue (although I am sure it is partially related), but one concern I have is how the buffer is ultimately transferred to the SPI. In devices where the full buffer memory can be allocated (and, consequently, no convert-on-write is required) the buffer should be should be stored in such a way that `write_array()` may be used to perform the transfer. We should avoid "spoon feeding" (see https://github.com/esphome/esphome/blob/c12c9e97c2574896ee71b4c686202cb9ccdc86d4/esphome/components/ili9341/ili9341_display.cpp#L100) for an example) the SPI when conversion is not absolutely required; as I understand it, `write_array()` uses some DMA so the transfer is faster and does not tie up the processor for so long. I understand that this is unavoidable when, to conserve memory, data in the buffer is in a format that differs from what the display requires; however, when it _is_ possible, we should make sure to do so. This is probably more up to the individual components than the buffer mechanism itself. Still, I think it's something to keep in mind if we are concerned about performance.

File: tests/test1.yaml (hunk 6 above)
Comment (kbx81) [reply tests/test1.yaml:1720]: Second this. 🙂

Comment (SenexCrenshaw) [reply tests/test1.yaml:1720]: Will do

**Comment by SenexCrenshaw** [comment]
@kbx81 Yes! I agree. Not using write-array bothered me too but i didnt dig into yet. I should have noted that. I couldnt do use it as of now since it with the 16bit buffer because write-array is 8bit only. A cursory glance looks like fastspi_arm_sam.h is used for the hardware spi transfer and it actually has a note about optimizing for 16bit words instead of 8bits. I will look into this and see if we can take advantage of 16bit hardware transfer. I will look to create a 16bit write_array that will handle software or hardware depending on the MCU.

**Comment by SenexCrenshaw** [comment]
Ill clean up the recent push. I am working out a partial update scheme and testing some different methods.

**Comment by SenexCrenshaw** [comment]
I added in a partial update method like in #1233 . However I did notice fill gets called on every update which negates how it was done in that PR. I changed displaybuffer to support a display_clear method that can be overwritten. As is it just works like before but for partial updates I override it to do nothing... 

I was thinking about the naming. It would be great if displaybuffer was named display and this became displaybuffer... 

@glmnet  I didnt add in the optional piece yet since im not the best with the python api. I am researching but not getting very far. Im looking for an example or something if there are any.

**Comment by SenexCrenshaw** [comment]
Made buffer optional. I am not sure how to include integrations besides using AUTO_LOAD. So I did this for the default. The problem is if one uses a different buffer the AUTO_LOAD (the default buffer type) will still be copied. This adds a 1388 byte overhead. Not awful but there might be better ways to do this.

**Comment by SenexCrenshaw** [comment]
Rebased with the latest color changes

**Comment by jesserockz** [comment]
@SenexCrenshaw Whats happening with this PR?

**Comment by SenexCrenshaw** [comment]
> @SenexCrenshaw Whats happening with this PR?

This started as a POC to swap out different buffer storage types. @glmnet and I went back and forth playing around with it. I think it has good bones and would help keeping the various display drivers API and configuration consistent as well as share any useful tricks (efficient storage, palleted colors, partial updates, PSRAM...) but I don't have a ton of time to work on this. I think if this is something that could be useful we would have to agree on a direction with some input from folks besides me for sure. 

If you think there is value in this I could pick it back up. If not it could be closed.  LMK

**Pull request closed without merging**